### PR TITLE
Add new revision of crkbd

### PR DESCRIPTION
**Pull request by foostan**
## Description

- Add the new revision of the crkbd for [Corne Version 4](https://github.com/foostan/crkbd/pull/208)
- The new version uses RP2040 and has some fixes

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/crkbd/rev1/rules.mk
Comment (waffle87) [inline keyboards/crkbd/rev1/rules.mk:2]:
```suggestion
# This file intentionally left blank
```

File: keyboards/crkbd/rev2/config.h
Comment (waffle87) [inline keyboards/crkbd/rev2/config.h:34]:
```suggestion
#define I2C_DRIVER I2CD1
#define OLED_UPDATE_INTERVAL 50
```

File: keyboards/crkbd/rev2/crkbd.h
Comment (waffle87) [inline keyboards/crkbd/rev2/crkbd.h:1]: The matrix data must be migrated to `info.json` and this file removed.

File: keyboards/crkbd/rev2/crkbd.json
Comment (waffle87) [inline keyboards/crkbd/rev2/crkbd.json:1]: This file should be removed.

File: keyboards/crkbd/rev2/info.json
Comment (waffle87) [inline keyboards/crkbd/rev2/info.json:16]: These are all already enabled at the keyboard level `info.json`.
```suggestion
        "encoder": true,
        "rgb_matrix": true
```

File: keyboards/crkbd/rev2/info.json
Comment (waffle87) [inline keyboards/crkbd/rev2/info.json:18]: No diodes in direct pin configuration?
```suggestion
```

File: keyboards/crkbd/rev2/info.json
Comment (waffle87) [inline keyboards/crkbd/rev2/info.json:114]:
```suggestion
        "driver": "vendor",
        "pin": "GP10"
```

File: keyboards/crkbd/rev2/info.json
Comment (waffle87) [inline keyboards/crkbd/rev2/info.json:52]:
```suggestion
        },
        "transport": {
            "sync": {
                "matrix_state": true,
                "layer_state": true,
                "modifiers": true
            }
        }
```

File: keyboards/crkbd/rev2/rules.mk
Comment (waffle87) [inline keyboards/crkbd/rev2/rules.mk:2]: This feature cannot be enabled at the keyboard level.
```suggestion
```

File: keyboards/crkbd/rev2/rules.mk
Comment (waffle87) [inline keyboards/crkbd/rev2/rules.mk:3]: This define needs to be placed in a proper `halconf.h` file:
```c
// Copyright 2023 QMK
// SPDX-License-Identifier: GPL-2.0-or-later
#pragma once

#define HAL_USE_I2C TRUE

#include_next <halconf.h>
```
```suggestion
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/crkbd/keymaps/default/keymap.c
Comment (zvecr) [inline keyboards/crkbd/keymaps/default/keymap.c:22]: You cannot make this change as it makes the "generic" keymap fail on all the other revisions as they do not support `LAYOUT_split_3x6_3_ex2`. Keymaps would have to be moved to be per revision, or a `LAYOUT_split_3x6_3` implementation added to rev2 so this keymap compiles without changes on all revisions.

File: keyboards/crkbd/keymaps/default/keymap.c
Comment (zvecr) [inline keyboards/crkbd/keymaps/default/keymap.c:76]: Same goes here, encoders are not enabled on other revisions and this will fail. It should be surrounded by `#ifdef ENCODER_MAP_ENABLE` guards.

File: keyboards/crkbd/keymaps/via/rules.mk
Comment (zvecr) [inline keyboards/crkbd/keymaps/via/rules.mk:1]: This is likely required to keep the firmware size down on older revisions.

File: keyboards/crkbd/rules.mk
Comment (zvecr) [inline keyboards/crkbd/rules.mk:5]: Changing the default folder would have to go via `develop`.

**Review by zvecr — CHANGES_REQUESTED** [review]

**Comment by foostan** [comment]
Thank you for reviewing quickly! I’ll fix and confirm on the actual board.

File: keyboards/crkbd/rev2/mcuconf.h
Comment (sigprof) [inline keyboards/crkbd/rev2/mcuconf.h:23]:
```suggestion
```
Apparently you are not using I2C0 anywhere.

File: keyboards/crkbd/rev2/info.json
Comment (sigprof) [inline keyboards/crkbd/rev2/info.json:206]: This looks like an unfortunate PCB misdesign — break-off keys (`[0, 0]`, `[1, 0]`, `[2, 0]`) should have been last in the chain.  As implemented now, breaking off the outer column would also disable the LEDs under the extra inner keys (and you would also be relying on the behavior of LEDs with floating DIN inputs).

File: keyboards/crkbd/rev2/info.json
Comment (sigprof) [inline keyboards/crkbd/rev2/info.json:55]: Consider adding `LAYOUT_split_3x5_3` and `LAYOUT_split_3x6_3` here, so that the keymaps made for the previous versions could also work on this version without changes.  This would also be useful for people who want to use OLEDs instead of extra keys on the new version.

**Review by sigprof — CHANGES_REQUESTED** [review]

File: keyboards/crkbd/rev2/info.json
Comment (Cipulot) [reply keyboards/crkbd/rev2/info.json:206]: @sigprof this aspect actually seems to be accounted in the PCB design by including s solder bridge that allows the redirection of the DIN/DOUT signals for the LED chain.

Comment (sigprof) [reply keyboards/crkbd/rev2/info.json:206]: The solder bridge solves the hardware part of the problem; however, after closing that bridge the LEDs for the extra inner keys would have wrong indices (not matching the RGB matrix configuration), so the per-key lighting functionality for those keys would be broken unless you patch `g_led_config` appropriately.  Any indicator code would also need changes to take the changed indices into account.

Currently QMK does not offer any good solution in that area apart from defining separate sub-revisions of the PCB with separate `rgb_matrix` configs in `info.json`.

Comment (Cipulot) [reply keyboards/crkbd/rev2/info.json:206]: Indeed, I guess that the indices issue wasn't considered or simply not regarded as an important aspect. Ofc effects would look weird and all over the place.

File: keyboards/crkbd/keymaps/default/keymap.c
Comment (foostan) [reply keyboards/crkbd/keymaps/default/keymap.c]: Add `LAYOUT_split_3x6_3` to rev2
https://github.com/qmk/qmk_firmware/pull/22676/commits/ac38e912877e1b78fe7a712412330c3d3a6df71e

File: keyboards/crkbd/keymaps/default/keymap.c
Comment (foostan) [reply keyboards/crkbd/keymaps/default/keymap.c]: Add `#ifdef ENCODER_MAP_ENABLE` https://github.com/qmk/qmk_firmware/commit/ac38e912877e1b78fe7a712412330c3d3a6df71e

File: keyboards/crkbd/keymaps/via/rules.mk
Comment (foostan) [reply keyboards/crkbd/keymaps/via/rules.mk:1]: disable mousekey at the json file https://github.com/qmk/qmk_firmware/pull/22676/commits/e70e4b2ecbddcb9cc8d2d92d365bcfe75522c0a4

File: keyboards/crkbd/rules.mk
Comment (foostan) [reply keyboards/crkbd/rules.mk]: revert https://github.com/qmk/qmk_firmware/pull/22676/commits/e08935278640555270da85fef45fc68328d0a9c1.
I'll change the default folder after developing rev2 and create a PR to `develop` fist.

File: keyboards/crkbd/rev2/config.h
Comment (foostan) [reply keyboards/crkbd/rev2/config.h]: Where are deleted parameters set? It does not work well if I just delete them at least.

Comment (foostan) [reply keyboards/crkbd/rev2/config.h]: Ah, this one https://github.com/qmk/qmk_firmware/pull/22676#discussion_r1428670937. Thanks!

File: keyboards/crkbd/rev2/config.h
Comment (drashna) [inline keyboards/crkbd/rev2/config.h:1]: Honestly, given that the are different versions of the PCB, this should be named rev4.  Otherwise, there will likely be confusion and people attempting to flash the wrong firmware. 

We already get a lot of questions about the rev3 pcb and what version should be used.

File: keyboards/crkbd/rev2/crkbd.json
Comment (drashna) [inline keyboards/crkbd/rev2/crkbd.json:1]: This looks like it's a via JSON, and doesn't belong in this repo, and should be removed.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/crkbd/rev2/crkbd.h
Comment (foostan) [reply keyboards/crkbd/rev2/crkbd.h:1]: deleted

File: keyboards/crkbd/rev2/crkbd.json
Comment (foostan) [reply keyboards/crkbd/rev2/crkbd.json:1]: deleted

File: keyboards/crkbd/rev2/info.json
Comment (foostan) [reply keyboards/crkbd/rev2/info.json:259]: hmm, you are correct. I thought I could just provide a separately modified and built file for 3x5, but it would be better if it could be expressed in info.json, which it is.

File: keyboards/crkbd/rev2/config.h
Comment (foostan) [reply keyboards/crkbd/rev2/config.h:1]: Hmm, that is a difficult issue. I'm pretty sure the keyboard version doesn't match the firmware version. However, if the firmware can be shared even if the keyboard version is increased, there is a thought that it should be so.

Would rev4 be better? Or I would like to make it v4.

File: keyboards/crkbd/rev2/crkbd.json
Comment (foostan) [reply keyboards/crkbd/rev2/crkbd.json:1]: I already delete, thanks!

File: keyboards/crkbd/rev2/info.json
Comment (foostan) [reply keyboards/crkbd/rev2/info.json]: Enable `LAYOUT_split_3x5_3` and `LAYOUT_split_3x6_3` at rev2

File: keyboards/crkbd/rev2/info.json
Comment (foostan) [reply keyboards/crkbd/rev2/info.json:257]: How about separating the revision for `rev2_mini` since the number of keys and LEDs are different?

File: keyboards/crkbd/rev2/info.json
Comment (dunk2k) [inline keyboards/crkbd/rev2/info.json:26]: already defined in `crkbd/info.json`
```suggestion
```

File: keyboards/crkbd/info.json
Comment (dunk2k) [inline keyboards/crkbd/info.json:16]: enabled in `build.lto`
```suggestion
```

**Review by dunk2k — CHANGES_REQUESTED** [review]

File: keyboards/crkbd/keymaps/via/rules.mk
Comment (dunk2k) [inline keyboards/crkbd/keymaps/via/rules.mk:2]: already enabled with `build.lto` in `crkbd/info.json`
```suggestion
```

**Review by dunk2k — CHANGES_REQUESTED** [review]

File: keyboards/crkbd/rev2/config.h
Comment (foostan) [reply keyboards/crkbd/rev2/config.h:1]: I updated the name of new revision to `rev4`, so could you confirm it.

File: keyboards/crkbd/rev2/info.json
Comment (foostan) [reply keyboards/crkbd/rev2/info.json]: I added `rev4_mini`. How about it?

File: keyboards/crkbd/rev2/config.h
Comment (drashna) [reply keyboards/crkbd/rev2/config.h:1]: Either seems appropriate, but I'd prefer keeping the naming consistent within the repo (eg, rev4 vs v4/ver4).  Though, in either case, it would be worth adding to the readme about what version to use for which pcb version.

Comment (foostan) [reply keyboards/crkbd/rev2/config.h:1]: I updated readme https://github.com/qmk/qmk_firmware/pull/22676/commits/90534aba7469467513c564d36ac6c81e1dd191fa

**Comment by foostan** [comment]
@sigprof @drashna @zvecr @waffle87 Thank you for your review. I fixed the reviewed points so could you confirm them again?

**Review by drashna — APPROVED** [review]

**Review by fauxpark — CHANGES_REQUESTED** [review]
Keyboard updates need to go through `develop`.

**Comment by foostan** [comment]
@fauxpark I changed the base branch from master to develop and the CI result looks good.

**Comment by foostan** [comment]
@drashna  You already approved this PR, so could you merge this PR? or Should I wait for approval from other members?  I can't get it re-reviewed anytime soon from other members...

**Comment by foostan** [comment]
Hi, @fauxpark @sigprof @zvecr @waffle87 Do you have any comments on the changes to your review?
@drashna Could you let me know if there is anything else I can do to merge this PR?

**Comment by foostan** [comment]
Since it takes time to change the keyboard settings, I will manage it separately from the core. Thank you for reviewing.

**Pull request closed without merging**
